### DISCUSSION TITLE
[Feature] Get rid of remote run-time dependencies

### DISCUSSION THREAD

**Discussion by mradalbert** [immich-app/immich#4909] [Feature Request]
### The feature

I tried to isolate Immich docker containers from the internet (docker internal networks) as a security measure. Unfortunately some containers print warning and error logs due to lost internet connectivity:

- `WARN [ServerInfoService] Unable to run version check: Error: getaddrinfo EAI_AGAIN api.github.com` from `server` container, not very important, I guess,
- `machine-learning` container seems to try to download ML models, and periodically prints massive error trace.

### Proposed solutions:
- incorporate ML models into container image or
- separate models acquisition to a simple, isolated and transparent container or
- add option to use local models without loading them from remote (possibly upgrade docker stack to build new layer with downloaded models during deployment).
- ideally modify docker stack to define default network as `internal`.
 
It may look like paranoia, and maybe it is :smile:, but Immich has access to very personal data. Taking care of this may be good from security standpoint and would improve sense of trust and security for users.

### Platform

- [X] Server
- [ ] Web
- [ ] Mobile

**Comment by bo0tzz** [comment]
You can download the models manually and place them in the ml container's cache volume.

**Reply by mradalbert** [reply]
I can let immich download models and switch to internal networking post factum, which is even more convenient. Unfortunately I have just noticed that Immich is malfunctioning, when network is in internal mode: New photos are not processed after uploading from phone (no metadata, thumbnail, preview or even a filename in details).

**Reply by mradalbert** [reply]
Search is non-functional as well

**Reply by jrasm91** [reply]
Machine learning models are huge and different people use different ones, so downloading them on demand makes sense.

Also you can disable the version check in the admin settings

**Reply by bo0tzz** [reply]
Immich should definitely work without an internet connection after it's downloaded models and such. Are there any errors in the logs?

**Reply by mradalbert** [reply]
Prior to yesterday internet connectivity was disabled and I have impression that errors in ML container's logs were occurring (not 100% sure). Yesterday I have performed upgrade from 1.84 to 1.85 (problems were occurring), then enabled internet connectivity for default network (problems stopped). Now when internet connectivity is disabled Immich seems to work as intended. Unfortunately I didn't keep logs - I was naive enough to think that the problem is easily reproducible.

**Comment by mradalbert** [comment]
Ok, I have just updated Immich server to 1.88.2 and new photos backed up from Android do not have details in the web UI. On Android side photos are marked as synced. In the web photos are in following state:
![image](https://github.com/immich-app/immich/assets/26419317/4f00d618-e484-4e41-ab3a-1f7cc257fc8e)
Photos can be downloaded - only are not processed by server.

microservices container is crashing. It uses 1-2 CPUs, gets to about 2.2 GB of RAM (I have above 10 GB avaliable memory) and then crashes:
```<--- Last few GCs --->
[7:0x3ee7e310000]   106527 ms: Scavenge 2033.7 (2078.0) -> 2030.7 (2078.5) MB, 4.45 / 0.00 ms  (average mu = 0.220, current mu = 0.189) allocation failure; 
[7:0x3ee7e310000]   106552 ms: Scavenge 2034.7 (2079.0) -> 2031.7 (2079.5) MB, 4.56 / 0.00 ms  (average mu = 0.220, current mu = 0.189) allocation failure; 
[7:0x3ee7e310000]   106579 ms: Scavenge 2035.4 (2079.7) -> 2032.4 (2088.2) MB, 6.52 / 0.00 ms  (average mu = 0.220, current mu = 0.189) allocation failure; 
<--- JS stacktrace --->
FATAL ERROR: Ineffective mark-compacts near heap limit Allocation failed - JavaScript heap out of memory
 1: 0xc99960 node::Abort() [immich_microservices]
 2: 0xb6ffcb  [immich_microservices]
 3: 0xebe910 v8::Utils::ReportOOMFailure(v8::internal::Isolate*, char const*, v8::OOMDetails const&) [immich_microservices]
 4: 0xebebf7 v8::internal::V8::FatalProcessOutOfMemory(v8::internal::Isolate*, char const*, v8::OOMDetails const&) [immich_microservices]
 5: 0x10d06a5  [immich_microservices]
 6: 0x10d0c34 v8::internal::Heap::RecomputeLimits(v8::internal::GarbageCollector) [immich_microservices]
 7: 0x10e7b24 v8::internal::Heap::PerformGarbageCollection(v8::internal::GarbageCollector, v8::internal::GarbageCollectionReason, char const*) [immich_microservices]
 8: 0x10e833c v8::internal::Heap::CollectGarbage(v8::internal::AllocationSpace, v8::internal::GarbageCollectionReason, v8::GCCallbackFlags) [immich_microservices]
 9: 0x10ea49a v8::internal::Heap::HandleGCRequest() [immich_microservices]
10: 0x1055907 v8::internal::StackGuard::HandleInterrupts() [immich_microservices]
11: 0x14f7322 v8::internal::Runtime_StackGuardWithGap(int, unsigned long*, v8::internal::Isolate*) [immich_microservices]
12: 0x7fbcf0819ef6 
[Nest] 7  - 11/28/2023, 11:58:28 AM     LOG [NestFactory] Starting Nest application...
[Nest] 7  - 11/28/2023, 11:58:28 AM     LOG [InstanceLoader] TypeOrmModule dependencies initialized +28ms
[Nest] 7  - 11/28/2023, 11:58:28 AM     LOG [InstanceLoader] BullModule dependencies initialized +0ms
[Nest] 7  - 11/28/2023, 11:58:28 AM     LOG [InstanceLoader] ConfigHostModule dependencies initialized +1ms
[Nest] 7  - 11/28/2023, 11:58:28 AM     LOG [InstanceLoader] DiscoveryModule dependencies initialized +0ms
[Nest] 7  - 11/28/2023, 11:58:28 AM     LOG [InstanceLoader] ScheduleModule dependencies initialized +0ms
[Nest] 7  - 11/28/2023, 11:58:28 AM     LOG [InstanceLoader] ConfigModule dependencies initialized +5ms
[Nest] 7  - 11/28/2023, 11:58:28 AM     LOG [InstanceLoader] BullModule dependencies initialized +0ms
[Nest] 7  - 11/28/2023, 11:58:28 AM     LOG [InstanceLoader] BullModule dependencies initialized +0ms
[Nest] 7  - 11/28/2023, 11:58:28 AM     LOG [InstanceLoader] TypeOrmCoreModule dependencies initialized +143ms
[Nest] 7  - 11/28/2023, 11:58:28 AM     LOG [InstanceLoader] TypeOrmModule dependencies initialized +0ms
[Nest] 7  - 11/28/2023, 11:58:28 AM     LOG [InstanceLoader] InfraModule dependencies initialized +3ms
[Nest] 7  - 11/28/2023, 11:58:28 AM     LOG [InstanceLoader] DomainModule dependencies initialized +12ms
[Nest] 7  - 11/28/2023, 11:58:28 AM     LOG [InstanceLoader] MicroservicesModule dependencies initialized +0ms
```

**Reply by mradalbert** [reply]
Forced thumbnail generation in administration page, and thumbnails were generated

**Reply by mradalbert** [reply]
Metadata generation job fails

**Reply by mradalbert** [reply]
Job succeeded after disabling `internal` network mode for default network

**Reply by mradalbert** [reply]
re-enabled `internal` network mode and now both thumbnail generation and metadata extraction job (and microservices container) fail
